src/rgw/rgw_cache.h

```diff
@@ -26,6 +36,137 @@ enum {
 
 #define mydout(v) lsubdout(T::cct, rgw, v)
 
+/*engage1*/
+struct DataCache;
+class L2CacheThreadPool;
+class HttpL2Request;
+
+struct ChunkDataInfo : public LRUObject {
+	CephContext *cct;
+	uint64_t size;
+	time_t access_time;
+	string address;
+	string oid;
+	bool complete;
+	struct ChunkDataInfo *lru_prev;
+	struct ChunkDataInfo *lru_next;	
+
+	ChunkDataInfo(): size(0) {}
+
+	void set_ctx(CephContext *_cct) {
+		cct = _cct;
+	}
+
+	void dump(Formatter *f) const;
+	static void generate_test_instances(list<ChunkDataInfo*>& o);
+};
+
+struct cacheAioWriteRequest{
+	string oid;
+	void *data;
+	int fd;
+	struct aiocb *cb;
+	DataCache *priv_data;
+	CephContext *cct;
+	
+	cacheAioWriteRequest(CephContext *_cct) : cct(_cct) {}
+	int create_io(bufferlist& bl, unsigned int len, string oid);
+	
+	void release() {
+		::close(fd);
+		cb->aio_buf = NULL;
+		free(data);
+		data = NULL;
+		free(cb);
+		free(this);
+	}
+};
+
+struct DataCache {
+
+private:
+  std::map<string, ChunkDataInfo*> cache_map;
+  std::list<string> outstanding_write_list;
+  int index;
+  RWLock lock;
+  Mutex  cache_lock;
+  Mutex  req_lock;
+  Mutex  eviction_lock;
+  CephContext *cct;
+  enum _io_type { 
+    SYNC_IO = 1,
+    ASYNC_IO = 2,
+    SEND_FILE = 3
+  } io_type;
+
+  struct sigaction action;
+  long long free_data_cache_size;
+  long long outstanding_write_size;
+  L2CacheThreadPool *tp;
+  struct ChunkDataInfo *head;
+  struct ChunkDataInfo *tail;
+
+private:
+  void add_io();
+
+public:
+  DataCache();
+  ~DataCache() {}
+
+  bool get(string oid);
+  void put(bufferlist& bl, unsigned int len, string obj_key);
+  int io_write(bufferlist& bl, unsigned int len, std::string oid);
+  int create_aio_write_request(bufferlist& bl, unsigned int len, std::string oid);
+  void cache_aio_write_completion_cb(cacheAioWriteRequest *c);
+  size_t random_eviction();
+  size_t lru_eviction();
+  string hash_uri(string dest);
+  string deterministic_hash(string oid);
+  void remote_io(struct librados::L2CacheRequest *l2request); 
+  void init_l2_request_cb(librados::completion_t c, void *arg);
+  void push_l2_request(librados::L2CacheRequest*l2request);
+  void l2_http_request(off_t ofs , off_t len, std::string oid);
+  void init(CephContext *_cct) {
+    cct = _cct;
+    free_data_cache_size = cct->_conf->rgw_datacache_size;
+    head = NULL;
+    tail = NULL;
+  }
+
+  void lru_insert_head(struct ChunkDataInfo *o) {
```

Comment:
I think this is ok for now (hand-rolled LRU list), but something based on boost::intrusive would be preferable at some future point